File: Source/Swift/MDSDocumentBackingCache.swift
//
//  MDSDocumentBackingCache.swift
//  Mini Document Storage
//
//  Created by Stevo on 10/9/19.
//  Copyright © 2019 Stevo Brock. All rights reserved.
//

import Foundation

//----------------------------------------------------------------------------------------------------------------------
// MARK: MDSDocumentBackingCache
class MDSDocumentBackingCache<T : MDSDocumentBacking> {

	// MARK: Reference
	private class Reference {

		// MARK: Properties
		let	documentBacking :T

		var	lastReferencedDate :Date

		// MARK: Lifecycle methods
		//--------------------------------------------------------------------------------------------------------------
		init(documentBacking :T) {
			// Store
			self.documentBacking = documentBacking

			// Setup
			self.lastReferencedDate = Date()
		}

		// MARK: Instance methods
		//--------------------------------------------------------------------------------------------------------------
		func noteWasReferenced() { self.lastReferencedDate = Date() }
	}

	// MARK: Properties
	private	let	limit :Int
	private	let	lock = ReadPreferringReadWriteLock()

	private	var	referenceByDocumentID = [/* Document ID */ String : Reference]()

	// MARK: Lifecycle methods
	//------------------------------------------------------------------------------------------------------------------
	init(limit :Int = 1_000_000) {
		// Store
		self.limit = limit
	}

	// MARK: Instance methods
	//------------------------------------------------------------------------------------------------------------------
	func add(_ documentBackings :[T]) {
		// Update
		self.lock.write() {
			// Do all document backing infos
			documentBackings.forEach() { self.referenceByDocumentID[$0.documentID] = Reference(documentBacking: $0) }
		}
	}

	//------------------------------------------------------------------------------------------------------------------
	func documentBacking(for documentID :String) -> T? {
		// Return cached document, if available
		return self.lock.read() {
			// Retrieve
			if let reference = self.referenceByDocumentID[documentID] {
				// Note was referenced
				reference.noteWasReferenced()

				return reference.documentBacking
			} else {
				// Not found
				return nil
			}
		}
	}

	//------------------------------------------------------------------------------------------------------------------
	func queryDocumentIDs(_ documentIDs :[String]) -> (foundDocumentIDs :[String], notFoundDocumentIDs :[String]) {
		// Setup
		var	foundDocumentIDs = [String]()
		var	notFoundDocumentIDs = [String]()

		// Iterate document IDs
		self.lock.read() { documentIDs.forEach() {
			// Look up reference for this document ID
			if let reference = self.referenceByDocumentID[$0] {
				// Found
				foundDocumentIDs.append($0)
				reference.noteWasReferenced()
			} else {
				// Not found
				notFoundDocumentIDs.append($0)
			}
		} }

		return (foundDocumentIDs, notFoundDocumentIDs)
	}

	//------------------------------------------------------------------------------------------------------------------
	func queryDocumentBackings(_ documentIDs :[String]) -> (foundDocumentBackings :[T], notFoundDocumentIDs :[String]) {
		// Setup
		var	foundDocumentBackings = [T]()
		var	notFoundDocumentIDs = [String]()

		// Iterate document IDs
		self.lock.read() { documentIDs.forEach() {
			// Look up reference for this document ID
			if let reference = self.referenceByDocumentID[$0] {
				// Found
				foundDocumentBackings.append(reference.documentBacking)
				reference.noteWasReferenced()
			} else {
				// Not found
				notFoundDocumentIDs.append($0)
			}
		} }

		return (foundDocumentBackings, notFoundDocumentIDs)
	}

	//------------------------------------------------------------------------------------------------------------------
	func remove(_ documentIDs :[String]) {
		// Remove from map
		self.lock.write() {
			// Remove from storage
			documentIDs.forEach() { self.referenceByDocumentID[$0] = nil }
		}
	}
}

File: src/main/java/codes/stevobrock/minidocumentstorage/MDSDocumentBackingCache.kt
package codes.stevobrock.minidocumentstorage

import java.util.*
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.collections.ArrayList
import kotlin.concurrent.read
import kotlin.concurrent.timer
import kotlin.concurrent.write

//----------------------------------------------------------------------------------------------------------------------
class MDSDocumentBackingCache<T : Any> {

	// Types
	data class QueryDocumentIDsInfo(val foundDocumentIDs :List<String>, val notFoundDocumentIDs :List<String>)
	data class QueryDocumentBackingInfosInfo<T : Any>(val foundDocumentBackingInfos :List<MDSDocument.BackingInfo<T>>,
				val notFoundDocumentIDs :List<String>)

	// Reference
	//------------------------------------------------------------------------------------------------------------------
	private class Reference<T : Any> {

		// Properties
		val	documentBackingInfo :MDSDocument.BackingInfo<T>

		var	lastReferencedDate :Date

		// Lifecycle methods
		//--------------------------------------------------------------------------------------------------------------
		constructor(documentBackingInfo :MDSDocument.BackingInfo<T>) {
			// Store
			this.documentBackingInfo = documentBackingInfo

			// Setup
			this.lastReferencedDate = Date()
		}

		// Instance methods
		//--------------------------------------------------------------------------------------------------------------
		fun noteWasReferenced() { this.lastReferencedDate = Date() }
	}

	// Properties
	private	val limit :Int
	private	val lock = ReentrantReadWriteLock()

	private	var	referenceMap = HashMap</* Document ID */ String, Reference<T>>()
	private	var	timer :Timer? = null

	// Lifecycle methods
	//------------------------------------------------------------------------------------------------------------------
	constructor(limit :Int = 1000000) {
		// Store
		this.limit = limit
	}

	// Instance methods
	//------------------------------------------------------------------------------------------------------------------
	fun add(documentBackingInfos :List<MDSDocument.BackingInfo<T>>) {
		// Update
		this.lock.write() {
			// Do all the document backing infos
			documentBackingInfos.forEach() { this.referenceMap[it.documentID] = Reference((it)) }

			// Reset pruning timer if needed
			resetPruningTimerIfNeeded()
		}
	}

	//------------------------------------------------------------------------------------------------------------------
	fun documentBacking(documentID :String) :T? {
		// Return cached document, if available
		this.lock.read() {
			// Retrieve
			val	reference = this.referenceMap[documentID]
			if (reference != null) {
				// Note was referenced
				reference.noteWasReferenced()

				return reference.documentBackingInfo.documentBacking
			} else
				// Not found
				return null
		}
	}

	//------------------------------------------------------------------------------------------------------------------
	fun queryDocumentIDs(documentIDs :List<String>) :QueryDocumentIDsInfo {
		// Setup
		val	foundDocumentIDs = ArrayList<String>()
		val	notFoundDocumentIDs = ArrayList<String>()

		// Iterate document IDs
		this.lock.read() { documentIDs.forEach() {
			// Look up reference for this document ID
			val	reference = this.referenceMap[it]
			if (reference != null) {
				// Found
				foundDocumentIDs.add(it)
				reference.noteWasReferenced()
			} else
				// Not found
				notFoundDocumentIDs.add(it)
		} }

		return QueryDocumentIDsInfo(foundDocumentIDs, notFoundDocumentIDs)
	}

	//------------------------------------------------------------------------------------------------------------------
	fun queryDocumentBackingInfos(documentIDs :List<String>) :QueryDocumentBackingInfosInfo<T> {
		// Setup
		val foundDocumentInfos = ArrayList<MDSDocument.BackingInfo<T>>()
		val	notFoundDocumentIDs = ArrayList<String>()

		// Iterate document IDs
		this.lock.read() { documentIDs.forEach() {
			// Look up reference for this document ID
			val	reference = this.referenceMap[it]
			if (reference != null) {
				// Found
				foundDocumentInfos.add(reference.documentBackingInfo)
				reference.noteWasReferenced()
			} else
				// Not found
				notFoundDocumentIDs.add(it)
		} }

		return QueryDocumentBackingInfosInfo(foundDocumentInfos, notFoundDocumentIDs)
	}

	//------------------------------------------------------------------------------------------------------------------
	fun remove(documentIDs :List<String>) {
		// Remove from map
		this.lock.write() {
			// Remove from storage
			documentIDs.forEach() { this.referenceMap.remove(it) }

			// Reset pruning timer if needed
			resetPruningTimerIfNeeded()
		}
	}

	// Private methods
	//------------------------------------------------------------------------------------------------------------------
	private fun resetPruningTimerIfNeeded() {
		// Invalidate existing timer
		this.timer?.cancel()
		this.timer = null

		// Check if need to prune
		if (this.referenceMap.size > this.limit) {
			// Need to prune
			this.timer = timer(initialDelay = 5000.toLong(), period = 5000) {
				// Prune
				this@MDSDocumentBackingCache.lock.write() {
					// Only need to consider things if we have moved past the document limit
					val	countToRemove =
								this@MDSDocumentBackingCache.referenceMap.size - this@MDSDocumentBackingCache.limit
					if (countToRemove > 0) {
//						// Iterate all references
//						val	referencesToRemove = ArrayList<Reference<T>>()
//						var	earliestReferencedDate = Date
//						this@MDSDocumentBackingCache.referenceMap.values.forEach() {
//							// Compare date
//
//						}
					}
				}

				// Cleanup
				this@MDSDocumentBackingCache.timer = null
			}
		}
	}
}
